server/drafty/drafty.go

```diff
@@ -37,7 +93,137 @@ var tags = map[string]spanfmt{
 	"EX": {"", true},
 }
 
-// ToPlainText converts message payload from Drafy format to string.
+// Preview shortens Drafty to the specified length (in runes) and removes large content from entities making them
+// suitable for a one-line preview, for example for showing in push notifications.
+func Preview(content interface{}, length int) (interface{}, error) {
+	if content == nil {
+		return "", nil
+	}
+
+	var original map[string]interface{}
+
+	switch tmp := content.(type) {
+	case string:
+		return tmp, nil
+	case map[string]interface{}:
+		original = tmp
+	default:
+		return "", errUnrecognizedContent
+	}
+
+	txt, txtOK := original["txt"].(string)
+	fmt, fmtOK := original["fmt"].([]interface{})
+	ent, entOK := original["ent"].([]interface{})
+
+	// At least one must be set.
+	if !txtOK && !fmtOK && !entOK {
+		return "", errUnrecognizedContent
+	}
+
+	var textLen int
+	preview := make(map[string]interface{})
+	if txtOK {
+		runes := []rune(txt)
+		textLen = len(runes)
+		if textLen > length {
+			preview["txt"] = string(runes[:length])
+			textLen = length
+		} else {
+			preview["txt"] = txt
+		}
+		txt = preview["txt"].(string)
+	}
+
+	if len(fmt) > 0 {
+		// Old key to new key entity mapping.
+		entRefs := make(map[int]int)
+
+		// Count styles which start within the new length of the text and save entity keys as set.
+		for i := range fmt {
+			s := span{}
+			if err := s.fromMap(fmt[i]); err != nil {
+				return "", err
+			}
+
+			if s.at < textLen {
+				if s.tp == "" {
+					if s.key < 0 {
+						return "", errUnrecognizedContent
+					}
+
+					if _, ok := entRefs[s.key]; !ok {
+						entRefs[s.key] = len(entRefs)
+					}
+				}
+			}
+		}
+
+		// Allocate space for copying styles and entities.
+		var preview_fmt []map[string]interface{}
+		var preview_ent []map[string]interface{}
+		if len(entRefs) > 0 {
+			preview_ent = make([]map[string]interface{}, len(entRefs))
+		}
+		for i := range fmt {
+			old := span{}
+			// No need to check for errors: already checked earlier.
+			old.fromMap(fmt[i])
```

Comment:
You can save a `fromMap` call by storing the results from the same call on line 144 into an array.

Reply:
Done